confgenerator/testdata/valid/linux/logging-processor_exclude_logs/golden_fluent_bit_main.conf

```diff
@@ -0,0 +1,774 @@
+@SET buffers_dir=/var/lib/google-cloud-ops-agent/fluent-bit/buffers
+@SET logs_dir=/var/log/google-cloud-ops-agent/subagents
+
+[SERVICE]
+    Daemon                    off
+    Flush                     1
+    HTTP_Listen               0.0.0.0
+    HTTP_PORT                 2020
+    HTTP_Server               On
+    Log_Level                 info
+    storage.backlog.mem_limit 50M
+    storage.checksum          on
+    storage.max_chunks_up     128
+    storage.metrics           on
+    storage.sync              normal
+
+[INPUT]
+    Buffer_Chunk_Size 512k
+    Buffer_Max_Size   5M
+    DB                ${buffers_dir}/default_pipeline_syslog
+    Key               message
+    Mem_Buf_Limit     10M
+    Name              tail
+    Path              /var/log/messages,/var/log/syslog
+    Read_from_Head    True
+    Rotate_Wait       30
+    Skip_Long_Lines   On
+    Tag               default_pipeline.syslog
+    storage.type      filesystem
+
+[INPUT]
+    Buffer_Chunk_Size 512k
+    Buffer_Max_Size   5M
+    DB                ${buffers_dir}/p1_sample_logs
+    Key               message
+    Mem_Buf_Limit     10M
+    Name              tail
+    Path              /tmp/*.log
+    Read_from_Head    True
+    Rotate_Wait       30
+    Skip_Long_Lines   On
+    Tag               p1.sample_logs
+    storage.type      filesystem
+
+[INPUT]
+    Buffer_Chunk_Size 512k
+    Buffer_Max_Size   5M
+    DB                ${buffers_dir}/ops-agent-fluent-bit
+    Key               message
+    Mem_Buf_Limit     10M
+    Name              tail
+    Path              ${logs_dir}/logging-module.log
+    Read_from_Head    True
+    Rotate_Wait       30
+    Skip_Long_Lines   On
+    Tag               ops-agent-fluent-bit
+    storage.type      filesystem
+
+[FILTER]
+    Add   logging.googleapis.com/logName syslog
+    Match default_pipeline.syslog
+    Name  modify
+
+[FILTER]
+    Match      p1.sample_logs
+    Name       nest
+    Nest_under record
+    Operation  nest
+    Wildcard   *
+
+[FILTER]
+    Condition Key_value_matches $record['message'] (?i)^bar$
+    Match     p1.sample_logs
+    Name      modify
+    Set       __match_p1_sample_logs 1
+
+[FILTER]
+    Exclude __match_p1_sample_logs 1
+    Match   p1.sample_logs
+    Name    grep
+
+[FILTER]
+    Match           p1.sample_logs
+    Name            modify
+    Remove_wildcard __match_p1_sample_logs
+
+[FILTER]
+    Match        p1.sample_logs
+    Name         nest
+    Nested_under record
+    Operation    lift
+
```

Comment:
I guess i miss some context in here. Maybe it's mentioned in the design doc. Pardon me then.

Why it's being done this way, instead of just:
```

[FILTER]
    Exclude message ^ba$
    Match   p1.sample_logs
    Name    grep

```

Reply:
There are a few nuances with the `grep` filter that make things more complicated:

1) The `grep` filter (and any other filter that uses record accessors) does not let you use symbols in the root part of the path, so it would not be possible to filter on e.g. `logging.googleapis.com/severity`. But it is allowed to use symbols in non-root paths. So we `nest` those fields into a temporary `record` field and `lift` them afterwards in order to allow access to special fields.

2) We use `modify` to allow for AND, NOT, and grouping with parentheses. With `grep` filters alone you can only do a single layer of OR filters.

3) We use `(?i)^bar$` instead of `^bar$` to more closely match Cloud Logging queries which are case-insensitive when using `=` or `!=`.

Reply:
Thanks for the explanation.

One doubt about the point 1.  Below is a config that excludes syslogs that contains "systemd" (verified it works). I am curious what cases that you encountered that didn't work.
```
[sophieyfang_google_com@centos8 ~]$ sudo vim syslog.conf

[SERVICE]
    Daemon                    off
    Flush                     1
    HTTP_Listen               0.0.0.0
    HTTP_PORT                 2020
    HTTP_Server               On
    Log_Level                 debug
    storage.backlog.mem_limit 50M
    storage.checksum          on
    storage.max_chunks_up     128
    storage.metrics           on
    storage.sync              normal
[INPUT]
    Buffer_Chunk_Size 512k
    Buffer_Max_Size   5M
    Key               message
    Mem_Buf_Limit     10M
    Name              tail
    Path              /var/log/messages,/var/log/syslog
    Read_from_Head    True
    Rotate_Wait       30
    Skip_Long_Lines   On
    Tag               default_pipeline.syslog
[FILTER]
    name   grep
    match  default_pipeline.syslog
    exclude  message .*systemd.*
[OUTPUT]
    name  stdout
    match *
```

Reply:
The restriction applies to the path part which is `message` in that example, which doesn't have symbols, so it works fine. The following variation would not work because of the `.` and `/` symbols:

```
[FILTER]
    name   grep
    match  default_pipeline.syslog
    exclude  logging.googleapis.com/severity .*systemd.*
```

Reply:
I feel like we shall make the grep filter behaves correctly. 

- If the key doesn't start with '$', it shall be treated as a string. In this case, `logging.googleapis.com/severity`
- If the key starts with $, e.g. `$kubernetes['labels']['app']`, then it's a nested record key.

fluentbit code that gets the field(key):
https://github.com/fluent/fluent-bit/blob/master/plugins/filter_grep/grep.c#L101

Reply:
What about the case of special fields + nested fields? The user could want `httpRequest.requestUrl`. Should that look like:

```
    Exclude $logging.googleapis.com/http_request['requestUrl'] pattern
```

or

```
    Exclude $['logging.googleapis.com/http_request']['requestUrl'] pattern
```

or something else? IMO we should always use the `$` style even if the key has no symbols or nested keys because it keeps the ops agent code simpler.

If you're proposing that we make a change to Fluent Bit to avoid nest/lift then I would much prefer the second option above. It would need to apply to record accessors in general, not just for `grep`, since we still need `modify` to handle the boolean logic.

But at any rate I'm not sure if this change needs to happen now since it works fine with the nest/lift workaround. Unless there's a significant performance bottleneck caused by the nest/lift then I'm hesitant to block this feature on upstream work. We can always put in a feature request for FB and change ops agent in a later release.

Reply:
During the grep filter, it first prefixes the key with $, then it goes into the record accessor, and record accessor claims to support: https://github.com/fluent/fluent-bit/blob/master/src/flb_record_accessor.c#L65
![image](https://user-images.githubusercontent.com/8354694/143294130-7797e793-017b-44eb-8797-c483145cff85.png)
I wonder if  we can or shall fit `logging.googleapis.com/severity` as `key` into the case bucket of `$key`

Reply:
They don't document the allowable characters for `key` though, and from looking at the grammar it appears they only support alphanumeric, underscore and dot characters:

[parser](https://github.com/fluent/fluent-bit/blob/master/src/record_accessor/ra.y):

> ```
> /* Parse record accessor string: $key, $key['x'], $key['x'][...] */
> record_accessor: record_key
>       record_key:
>                   '$' IDENTIFIER
>                   {
>                     void *key;
> 
>                     rp->type = FLB_RA_PARSER_KEYMAP;
>                     key = flb_ra_parser_key_add(rp, $2);
>                     if (key) {
>                       rp->key = key;
>                     }
>                     flb_free($2);
>                   }
>                   |
>                   '$' IDENTIFIER record_subkey
>                   {
>                     void *key;
>                     rp->type = FLB_RA_PARSER_KEYMAP;
>                     key = flb_ra_parser_key_add(rp, $2);
>                     if (key) {
>                       rp->key = key;
>                     }
>                     flb_free($2);
>                   }
> ```

[lexer](https://github.com/fluent/fluent-bit/blob/master/src/record_accessor/ra.l):

> ```
> \'([^']|'{2})*\'           { yylval->string = remove_dup_quotes(yytext + 1, yyleng - 2); return STRING; }
> [_A-Za-z][A-Za-z0-9_.]*	   { yylval->string = flb_strdup(yytext); return IDENTIFIER; }
> ```

Although in my testing it doesn't actually work with `.` so there's a bug there anyhow. I suppose it would be sufficient to (1) add `/` to the allowed characters in the grammar, and (2) fix the `.` not working, since those are the only other symbols we need at the moment. But I still think it makes more sense to allow `'$' '[' STRING ']'` as a parser rule which can already accept almost any character.

Reply:
I will vote for `Exclude $['logging.googleapis.com/http_request']['requestUrl']` too.

Reply:
Created an internal tracking ticket for this.